staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_width.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"math"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+func newListWidthEstimator(countFn objectCountGetterFunc) WidthEstimatorFunc {
+	const (
+		// the minimum number of seats a list request occupies
+		minimumSeats uint = 3
+	)
+
+	return func(r *http.Request) Width {
+		requestInfo, ok := apirequest.RequestInfoFrom(r.Context())
+		if !ok {
+			return Width{Seats: minimumSeats}
+		}
+
+		query := r.URL.Query()
+		count := countFn.Get(key(requestInfo))
+
+		var fieldSelectorPresent bool
+		if len(query.Get("fieldSelector")) > 0 {
+			fieldSelectorPresent = true
+		}
+
+		var resourceVersionNotSet bool
+		if version, ok := query["resourceVersion"]; ok && len(version) > 0 && len(version[0]) == 0 {
+			resourceVersionNotSet = true
+		}
+
+		limit := parseLimit(query)
+
+		var estimatedObjectsToBeProcessed int64
+		switch {
+		case fieldSelectorPresent && count > 0:
+			estimatedObjectsToBeProcessed = count
+		case resourceVersionNotSet && count > 0:
+			estimatedObjectsToBeProcessed = count
+		case limit > 0:
+			estimatedObjectsToBeProcessed = limit
+		}
+
+		// for now, our rough estimate is to allocate one seat to each 100 obejcts that
+		// will be processed by the list request.
+		// we will come up with a different formula for the transformation function and/or
+		// fine tune this number in future iteratons.
+		seats := uint(math.Ceil(float64(estimatedObjectsToBeProcessed) / float64(100)))
+		if seats < minimumSeats {
```

Comment:
> For the smallest list the number of seats should actually be 1 (if there is just one object LIST is actually equally expensive to GET).

yes, i agree, 1-item list is equally expensive to GET, i was thinking of the list requests for which we simply return width=1 because we don't have enough context to estimate width. only a fraction of those list requests will be equally expensive to GET. on an average, a list is more expensive than a GET and may be we should set minimum to `2`, thoughts?